### DISCUSSION TITLE
32 bit raspberry pi limitations

### DISCUSSION THREAD

**Discussion by pizzadoglover911** [markqvist/Reticulum#262] [General]
My quest for an inexpensive pair of pi zeros to run reticulum has ended with compatibility issues. 32bit architecture won't support some dependencies.

If only I read that portion of the documentation _before_ ordering these pi zeros. 

Anywho pi zero w v2 is 64 bit so onto the next. Hopefully they fix their supply chain soon, prices and availability are poor.

Good day!

**Comment by gyv4** [comment]
Hi,

I think that there are no compatibility issues, it is just that some packages do not come pre-compiled any longer, and one has to compile them locally. Cryptography, if I remember correctly.
The pizeroW can still be used for reticulum.

**Reply by markqvist** [reply]
Yeah, that's correct. Last time I checked, at least, there were no pre-compiled packages for `cryptography` on 32-bit RasPi OS in pip. You can try `sudo apt install python3-cryptography` (which may also lack precompiled 32-bit packages at this point), or simply compiling it from source. It's a little cumbersone.

As a last resort, you can run pure-python Reticulum, without relying on PyCa/OpenSSL packages for Reticulum, by just forcing the pip install of `rns`, or installing `rnspure`. If you do this, please understand [this](https://markqvist.github.io/Reticulum/manual/understanding.html#cryptographic-primitives).

**Comment by gyv4** [comment]
To install reticulum on the 32 bit Raspbian GNU/Linux (tested on version 10 and 11), you'll need to install a couple of packages
`apt install libssl-dev python3-dev` 
and then a rust compiler. If you follow the instructions from https://www.rust-lang.org/tools/install you may end up with the installation failing because `rustup` tries to allocate a lot of memory while installing, and it will exceed the 512MB of a pizero or pizero2.
A workaround is to  set an environment variable
`export RUSTUP_IO_THREADS=1`
before running the `rustup` command. Apparently the single threaded version is less memory greedy.
Then you can proceed with the `pip install ... `usual instructions.
When it's time to install cryptography, it will take quite a long time to compile (on the pizeros), but it should work.

**Reply by markqvist** [reply]
Thanks a lot for these pointers!

**Reply by Swissbandit** [reply]
An easy fix would be,using a more powerful pi to prepare the image for the less powerful machine to use after the software is installed.
